### PR TITLE
automatically set LHOST

### PR DESCRIPTION
**Pull request by Auxilus**
this PR adds a new option `AutoLHOST` which if set and if LHOST is nil, will get the IP from Socket.ip_address_list.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use exploit/multi/handler`
- [ ] `set payload <payload>`
- [ ] `set AutoLHOST true`
- [ ] *set LHOST to random IP* 
- [ ] **verify** that it sets currect LHOST and you get a session with that

### CODE REVIEW

File: lib/msf/core/opt.rb [hunk 1]
```diff
@@ -26,7 +26,7 @@ def self.CPORT(default=nil, required=false, desc="The local client port")
     end
 
     # @return [OptAddress]
-    def self.LHOST(default=nil, required=true, desc="The listen address")
+    def self.LHOST(default=nil, required=false, desc="The listen address")
```
Comment (Auxilus) [inline lib/msf/core/opt.rb]: any better workaround for this? will this break things?

Comment (jmartin-r7) [reply lib/msf/core/opt.rb]: I broke tests that are based in it being "true". :trollface: 

When doing anything that will impact core code like this please be sure to migrate a test database with `RAILS_ENV=test rake db:create db:migrate` and run `rake spec` to determine what tests might depend on known state.

**Comment by Auxilus** [comment]
Maybe use this to autodetect the LHOST if the given LHOST fails to bind? I think that way we wont need to change opt.rb

**Comment by Auxilus** [comment]
**before**

```
msf5 exploit(multi/handler) > set lhost 11.11.11.11
lhost => 11.11.11.11
msf5 exploit(multi/handler) > set autolhost false
autolhost => false
msf5 exploit(multi/handler) > exploit

[-] Handler failed to bind to 11.11.11.11:4444:-  -
[*] Started reverse TCP handler on 0.0.0.0:4444
```

**Comment by Auxilus** [comment]
**after**

```
msf5 exploit(multi/handler) > set autolhost true
autolhost => true
msf5 exploit(multi/handler) > exploit

[-] Privided LHOST failed to bind
[*] LHOST automatically set to 192.168.1.33
[*] Started reverse TCP handler on 192.168.1.33:4444
```

**Comment by Auxilus** [comment]
should I make `AutoLHOST` true by default?

**Comment by jmartin-r7** [comment]
Please rebase this on master and squash some of commits, 11 commits to change 18 lines in 1 file is just a bit too noisy.

**Pull request closed without merging**